pkg/errors/sdk_errors.go

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package errors
+
+// SDKError is an error type for Mainflux SDK.
+type SDKError interface {
+	Error
+	StatusCode() int
+}
+
+var _ SDKError = (*sdkError)(nil)
+
+type sdkError struct {
+	*customError
+	statusCode int
+}
+
+func (ce *sdkError) StatusCode() int {
+	return ce.statusCode
+}
+
+// NewSDKError returns an SDK Error that formats as the given text.
+func NewSDKError(text string) SDKError {
+	return &sdkError{
+		customError: &customError{
+			msg: text,
+			err: nil,
+		},
+	}
+}
+
+// NewSDKErrorWithStatus returns an SDK Error setting the status code.
+func NewSDKErrorWithStatus(msg string, statusCode int) SDKError {
+	return &sdkError{
+		statusCode: statusCode,
+		customError: &customError{
+			msg: msg,
+			err: nil,
+		},
+	}
+}
```

Comment:
Hello @AryanGodara , @dborovcanin 

@dborovcanin as you suggested this approach in CERTS Auto renewal PR, 
 Here by I'm proposing the same approach for all the SDK 

Code proposal for ```pkg/errors/sdk_error.go```

```go 

// Copyright (c) Mainflux
// SPDX-License-Identifier: Apache-2.0

package errors

import (
	"encoding/json"
	"io"
	"net/http"
)

var (
	errFailedToReadBody = New("failed to read http response body")
	errRespBodyNotJSON  = New("response body is not a valid JSON")
	errJSONKeyNotFound  = New("response body expected error message json key not found")
)

// SDKError is an error type for Mainflux SDK.
type SDKError interface {
	Error
	StatusCode() int
}

var _ SDKError = (*sdkError)(nil)

type sdkError struct {
	*customError
	statusCode int
}

func (ce *sdkError) StatusCode() int {
	return ce.statusCode
}

// NewSDKError returns an SDK Error that formats as the given text.
func NewSDKError(text string) SDKError {
	return &sdkError{
		customError: &customError{
			msg: text,
			err: nil,
		},
	}
}

// NewSDKErrorWithStatus returns an SDK Error setting the status code.
func NewSDKErrorWithStatus(msg string, statusCode int) SDKError {
	return &sdkError{
		statusCode: statusCode,
		customError: &customError{
			msg: msg,
			err: nil,
		},
	}
}

// CheckError will check for error in http response.
func CheckError(resp *http.Response, expectedStatusCodes ...int) error {
	for _, expectedStatusCode := range expectedStatusCodes {
		if resp.StatusCode == expectedStatusCode {
			return nil
		}
	}

	b, bErr := io.ReadAll(resp.Body)
	if bErr != nil {
		e := Wrap(errFailedToReadBody, bErr)
		return Wrap(NewSDKErrorWithStatus("", resp.StatusCode), e)
	}

	var content map[string]interface{}
	err := json.Unmarshal(b, &content)
	if err != nil {
		e := Wrap(errRespBodyNotJSON, New(string(b)))
		return Wrap(NewSDKErrorWithStatus("", resp.StatusCode), e)
	}

	if msg, ok := content["error"]; ok {
		if v, ok := msg.(string); ok {
			return NewSDKErrorWithStatus(v, resp.StatusCode)
		}
		return NewSDKErrorWithStatus("unknown error", resp.StatusCode)
	}
	e := Wrap(errJSONKeyNotFound, New(string(b)))
	return Wrap(NewSDKErrorWithStatus("", resp.StatusCode), e)
}
```

In the above code ```CheckError```  can be use in response of SDK, 


Example for at  ```pkg/sdk/go/bootstrap.go``` in  function ```AddBootstrap```

```go
func (sdk mfSDK) AddBootstrap(token string, cfg BootstrapConfig) (string, error) {
	data, err := json.Marshal(cfg)
	if err != nil {
		return "", err
	}

	url := fmt.Sprintf("%s/%s", sdk.bootstrapURL, configsEndpoint)
	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(data))
	if err != nil {
		return "", err
	}
	
	resp, err := sdk.sendRequest(req, token, string(CTJSON))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
        err = errors.CheckError(resp, http.StatusCreated ,   http.StatusOK)

	if err != nil {
		return "", err
	}
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	id := strings.TrimPrefix(resp.Header.Get("Location"), "/things/configs/")
	return id, nil
}

```


Example for  at ```pkg/sdk/go/things.go``` in  function ```UpdateThing```

```go
func (sdk mfSDK) UpdateThing(t Thing, token string) error {
	data, err := json.Marshal(t)
	if err != nil {
		return err
	}

	url := fmt.Sprintf("%s/%s/%s", sdk.thingsURL, thingsEndpoint, t.ID)

	req, err := http.NewRequest(http.MethodPut, url, bytes.NewReader(data))
	if err != nil {
		return err
	}

	resp, err := sdk.sendRequest(req, token, string(CTJSON))
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	
	return  errors.CheckError(resp,  http.StatusOK)
}

```

Reply:
Thanks for this @arvindh123 , I'll implement this in sdk/go and push the changes asap!

Reply:
> Let's make all the SDK methods (from `sdk.go` file) return SDKError, rather than `error`. When using SDK, we want to enable quick and simple access to the response details.

Done, changed from error interface to SDKError interface, in the sdk/go directory.